bypros/api.go

```diff
@@ -0,0 +1,213 @@
+	scID skipchain.SkipBlockID) error {
+
+	req := Follow{
+		Target: target,
+		ScID:   scID,
+	}
+
+	resp := EmptyReply{}
+
+	err := c.SendProtobuf(host, &req, &resp)
+	if err != nil {
+		return xerrors.Errorf("failed to send follow request: %v", err)
+	}
+
+	return nil
+}
+
+// UnFollow sends a request to stop following a node.
+func (c *Client) UnFollow(host *network.ServerIdentity) error {
+	req := UnFollow{}
+	resp := EmptyReply{}
+
+	err := c.SendProtobuf(host, &req, &resp)
+	if err != nil {
+		return xerrors.Errorf("failed to send follow request: %v", err)
+	}
+
+	return nil
+}
+
+// Query sends a query request. The query must be a read-only SQL query, for
+// example:
+//   Select * from cothority.block
+func (c *Client) Query(host *network.ServerIdentity, query string) ([]byte, error) {
+	req := Query{
+		Query: query,
+	}
+	resp := QueryReply{}
+
+	err := c.SendProtobuf(host, &req, &resp)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to send query request: %v", err)
+	}
+
+	return resp.Result, nil
+}
+
+// CatchUP sends a request to catch up from a particular block. A catch up will
+// update the database from the specified block until the end of the chain.
+// Every specifies the interval at which updates are sent.
+func (c *Client) CatchUP(ctx context.Context, host, target *network.ServerIdentity,
+	scID skipchain.SkipBlockID, fromBlock skipchain.SkipBlockID,
+	every int) (<-chan CatchUpResponse, error) {
+
+	req := CatchUpMsg{
+		ScID:        scID,
+		Target:      target,
+		FromBlock:   fromBlock,
+		UpdateEvery: every,
+	}
+
+	apiEndpoint, err := getWsAddr(req.Target)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get ws addr: %v", err)
+	}
+
+	apiURL := fmt.Sprintf("%s/%s/%s", apiEndpoint, ServiceName, "CatchUpMsg")
+
+	ws, err := c.OverlayClient.OpenWS(apiURL)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to open ws: %v", err)
+	}
+
+	buf, err := protobuf.Encode(&req)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to encode streaming request: %v", err)
+	}
+
+	err = ws.Write(websocket.BinaryMessage, buf)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to send streaming request: %v", err)
+	}
+
+	outChan := make(chan CatchUpResponse)
+
+	go func() {
+		defer func() {
+			err := ws.Write(websocket.CloseMessage, nil)
+			if err != nil {
+				log.Warnf("failed to send close: %v", err)
+			}
+			ws.Close()
+		}()
+
+		for {
+			_, buf, err = ws.Read()
+			if err != nil {
+				outChan <- CatchUpResponse{
+					Err: fmt.Sprintf("failed to read response: %v", err),
+				}
+				return
+			}
+
+			resp := CatchUpResponse{}
+
+			err = protobuf.Decode(buf, &resp)
+			if err != nil {
+				outChan <- CatchUpResponse{
+					Err: fmt.Sprintf("failed to decode response: %v", err),
+				}
+				return
+			}
+
+			outChan <- resp
+			if resp.Done {
+				close(outChan)
+				return
+			}
+		}
+	}()
+
+	return outChan, nil
+}
+
+// onetOverlay provides an overlay implementation based on onet and gorilla.
```

Comment:
Is this for easier testing?

Reply:
no, it is for proper decoupling. But it comes out that this makes life easier for testing ;)